### ISSUE TITLE
WASM support?

### ISSUE THREAD

**Issue by bluenote10** [ExpHP/npyz#65]
First of all, thanks for maintaining this library, this is quite awesome!

I was wondering if it would be feasible to officially support wasm? Ignoring zip support, it looks like it might be doable, but I'm not sure how much effort it really is. I made this minimal experiment:

```rust
fn main() {
    let data = include_bytes!("test.npy").as_slice();
    let npy = npyz::NpyFile::new(data).unwrap();
    println!("{} {:?}", npy.len(), npy.shape());
}
```

with a tiny `test.npy` generated via:

```py
import numpy as np
np.save("test.npy", np.array([1, 2, 3]))
```

Trying to compile/run it with `cargo run --target wasm32-unknown-unknown` currently fails to compile:

```
514 |     const SAFE_U16_CUTOFF: usize = 0x1_0000_0000 - 0x400;
    |                                    ^^^^^^^^^^^^^^^^^^^^^ attempt to compute `0_usize - 1024_usize`, which would overflow
```

Simply replacing the expression in line

https://github.com/ExpHP/npyz/blob/2e082784fd35a2e4d195fec64cf61d34d6392099/src/write.rs#L514

by `0xffff_fc00` seems to fix that issue and the compilation now succeeds.

Now there is only a runtime issue when trying to run the example:

```
target/wasm32-unknown-unknown/debug/data_roundtrip.wasm: 1: Syntax error: end of file unexpected
```

Do you think it is feasible to officially support wasm? I guess it mainly boils down to the question how many work-arounds for wasm's lack of unsigned integers are necessary to make it work?

**Comment by ExpHP** [comment]
Status: With #66, the following cargo features are supported/tested:

* [x] `derive`
* [ ] `arrayvec`
* [ ] `complex`
* [ ] `npz`

**Comment by ExpHP** [comment]
.....it just occurred to me that the reason it probably found no tests is because the lib isn't using the rust_bindgen_test attribute.  Oops.  Hang on.

**Comment by bluenote10** [comment]
> I would also need to make the module hierarchy public.

Yes, exactly that limitation feels pretty severe currently, and I also couldn't find any decent work-around. It should be noted that the [wasm book](https://rustwasm.github.io/docs/book/reference/add-wasm-support-to-crate.html) has a section on both adding wasm support and maintaining wasm support, and but they mainly reduce the problem to building for wasm. This suggests that a successful wasm build in combination with successful native tests may have a high chance of actually working at runtime on wasm as well? I have no clue how strong the "it compiles and it works for a different target" actually is, but I could imagine that in particular for a parsing library that operates on binary level, actual runtime checks would have been nice.

Thanks for the release in any case, and no worries regarding the delay (there's no hurry :wink:)!
